### PR TITLE
feat(expr): short-circuit optimization for coalesce

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

As title

## Refer to a related PR or issue link (optional)

#6203

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Will add some comments later

**Review by xiangjinwu — APPROVED** [review]

File: src/expr/src/expr/expr_coalesce.rs [hunk 1]
```diff
@@ -35,44 +36,46 @@ impl Expression for CoalesceExpression {
...
+                .as_ref()
+                .bitand(res_bitmap_ref)
+                .ones()
+                .for_each(|pos| {
+                    selection[pos] = Some(child_idx);
+                });
+            let res_vis: Vis = (!res_bitmap).into();
+            let new_vis = orig_vis & res_vis;
+            input.set_vis(new_vis);
+            children_array.push(res);
```
Comment (xxchan) [inline src/expr/src/expr/expr_coalesce.rs:59]: This logic is a little bit hard to follow 🤡

File: src/common/src/array/vis.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+            }
+            (VisRef::Compact(c1), VisRef::Compact(c2)) => {
+                assert_eq!(c1, c2);
+                Vis::Compact(c1)
+            }
+        }
+    }
+}
+
+impl<'a> std::ops::BitOr<Vis> for VisRef<'a> {
```
Comment (xxchan) [inline src/common/src/array/vis.rs:240]: Didn't fully understant the variants yet 😇. Could `impl<'a> std::ops::BitOr<Vis> for Vis` also saves a clone?

File: src/common/src/array/data_chunk.rs [hunk 3]
```diff
@@ -206,6 +140,13 @@ impl DataChunk {
         }
     }
 
+    pub fn set_vis(&mut self, vis: Vis) {
+        for column in &self.columns {
+            assert_eq!(vis.len(), column.array_ref().len())
+        }
+        self.vis2 = vis;
```
[the remark this answers is not in the document]
Comment (xiangjinwu) [reply src/common/src/array/data_chunk.rs:147]: A result of refactoring... We can call it `vis` (or even `visibility` if we are able to eliminate all `visibility: Bitmap` usage)

File: src/common/src/array/vis.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/common/src/array/vis.rs:240]: For Bitand, yes, but we don’t use it now, so we haven’t implemented it.

For Bitor, no.

File: src/common/src/buffer/bitmap.rs [hunk 4]
```diff
@@ -291,6 +291,13 @@ impl Bitmap {
         Bitmap::from_bytes_with_num_bits(bits, lhs.num_bits)
     }
 
+    pub fn ones(&self) -> impl Iterator<Item = usize> + '_ {
+        self.iter()
+            .enumerate()
+            .filter(|(_, bit)| *bit)
```
Comment (soundOfDestiny) [inline src/common/src/buffer/bitmap.rs:297]: what about using `filter_map` here?

Comment (TennyZhuang) [reply src/common/src/buffer/bitmap.rs:297]: IMO `filter` and `map` is more concise here.
